subsys/emul/i2c/emul_sbs_gauge.c

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Emulator for SBS 1.1 compliant smart battery fuel gauge.
+ * This emulator is *NOT* fully featured yet, only enough to get the sbs_gauge driver tests passing.
+ */
+
+#define DT_DRV_COMPAT sbs_sbs_gauge
+
+#define LOG_LEVEL CONFIG_SENSOR_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(sbs_sbs_gauge);
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/emul.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/i2c_emul.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/drivers/spi_emul.h>
+#include <zephyr/sys/byteorder.h>
+
+#include <sbs_gauge.h>
+
+/** Run-time data used by the emulator */
+struct sbs_gauge_emul_data {
+	/** sbs_gauge device being emulated */
+	const struct device *dev;
+	/** Current register to read (address) */
+	uint32_t cur_reg;
+};
+
+/** Static configuration for the emulator */
+struct sbs_gauge_emul_cfg {
+	/** I2C address of emulator */
+	uint16_t addr;
+};
+
+static void reg_write(const struct emul *target, int regn, int val)
+{
+	/* Unimplemented reg_write handler */
+	struct sbs_gauge_emul_data *data = target->data;
+
+	ARG_UNUSED(data);
+
+	LOG_INF("write %x = %x", regn, val);
+	switch (regn) {
+	default:
+		LOG_INF("Unknown write %x", regn);
+	}
+}
+
+static int reg_read(const struct emul *target, int regn)
+{
+	struct sbs_gauge_emul_data *data = target->data;
+	int val;
+
+	ARG_UNUSED(data);
+
+	LOG_INF("read %x =", regn);
+	switch (regn) {
+	case SBS_GAUGE_CMD_VOLTAGE:
+	case SBS_GAUGE_CMD_AVG_CURRENT:
+	case SBS_GAUGE_CMD_TEMP:
+	case SBS_GAUGE_CMD_ASOC:
+	case SBS_GAUGE_CMD_FULL_CAPACITY:
+	case SBS_GAUGE_CMD_REM_CAPACITY:
+	case SBS_GAUGE_CMD_NOM_CAPACITY:
+	case SBS_GAUGE_CMD_AVG_TIME2EMPTY:
+	case SBS_GAUGE_CMD_AVG_TIME2FULL:
+	case SBS_GAUGE_CMD_CYCLE_COUNT:
+	case SBS_GAUGE_CMD_DESIGN_VOLTAGE:
+		val = 1;
+		break;
+	default:
+		val = -EIO;
+		LOG_INF("Unknown read %x", regn);
```

Comment:
In this case, the log output will look a little weird. Things you could do instead:
* A single "read X = Y" `LOG_INF` after the `switch` while retaining this `LOG_INF`, producing 2 distinct log lines.
* Return early in `default:`, producing 1 log line in each case.
* A single "read X = Y" `LOG_INF` after the `switch`, formatted in a way that makes the `-EIO` obvious, eliminating the need for a separate error log. (I assume that the valid register values are something like 0x00 - 0xff.)

Reply:
Good catch. Thanks!

Reply:
I still see the "read X = Y" log line split across the "Unknown read X" log line.

Reply:
Ugh, I should make changes before 5pm. Thanks, again.